src/app/modules/shared_models/message_model.nim

```diff
@@ -275,37 +275,59 @@ QtObject:
       if(self.items[i].responseToMessageWithId == messageId):
         result.add(self.items[i].id)
 
-  proc findIndexBasedOnClockToInsertTo(self: Model, clock: int64, id: string): int =
-    for i in 0 ..< self.items.len:
-      if clock > self.items[i].clock:
-        return i
-      elif clock == self.items[i].clock: # break ties by message id
-        if id > self.items[i].id:
-          return i
-    return self.items.len
-
-  proc insertItemBasedOnClock*(self: Model, item: Item) =
-    if(self.findIndexForMessageId(item.id) != -1):
-      return
+  # sort predicate - most recent clocks first, break ties by message id
+  proc isGreaterThan(lhsItem, rhsItem: Item): bool =
+    return lhsItem.clock > rhsItem.clock or
+           (lhsItem.clock == rhsItem.clock and lhsItem.id > rhsItem.id)
 
+  proc insertItems(self: Model, position: int, items: seq[Item]) =
     let parentModelIndex = newQModelIndex()
     defer: parentModelIndex.delete
 
-    let position = self.findIndexBasedOnClockToInsertTo(item.clock, item.id)
-
-    self.beginInsertRows(parentModelIndex, position, position)
-    self.items.insert(item, position)
+    self.beginInsertRows(parentModelIndex, position, position + items.len - 1)
+    self.items.insert(items, position)
     self.endInsertRows()
 
     if position > 0:
       self.updateItemAtIndex(position - 1)
-    if position + 1 < self.items.len:
-      self.updateItemAtIndex(position + 1)
+    if position + items.len - 1 < self.items.len:
+      self.updateItemAtIndex(position + items.len)
     self.countChanged()
 
+  proc filterExistingItems(self: Model, items: seq[Item]): seq[Item] =
+    let existingItems = toHashSet(self.items.map(x => x.id))
+    return items.filter(item => not existingItems.contains(item.id))
+
   proc insertItemsBasedOnClock*(self: Model, items: seq[Item]) =
-    for item in items:
-      self.insertItemBasedOnClock(item)
+    # remove existing items and sort by most recent
+    let sortCmp = proc(lhs, rhs: Item): int =
+      return if isGreaterThan(lhs, rhs): -1 else: 1
+    let newItems = sorted(self.filterExistingItems(items), sortCmp)
+
+    # bulk insert algorithm, "two-pointer" technique
+    var current = 0
+    var new = 0
```

Comment:
A var named `new` is a bit scary because it's usually a reserved word. Also, it doesn't mean much, new what?

Reply:
Good point. `current` and `new` are pointers/indexes ("two-pointer" technique), it was obvious to me at the time of writing. Will update.

Reply:
Changed to `currentIdx` and `newIdx`.